### PR TITLE
Document end-to-end tests, load testing approach and SLO verification

### PR DESCRIPTION
**Pull request by vrutkovs**
Ref: OTA-182

/cc @steveeJ

### CODE REVIEW

**Comment by vrutkovs** [comment]
CentOS mirror failure

/retest

File: docs/developer/end-to-end-testing.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+Once the script is done it would start end-to-end tests from [e2e package](../e2e/tests). 
+
+First test (`e2e_channel_success`) checking the graph in the deployed Cincinnati and ensure that [expected releases](../../e2e/test/testdata) are present in the graph and have valid metadata. The test verifies that graph-builder is able to scrape production images from Quay correctly and policy-engine properly filters the graph by channel and architecture.
+
+### Load testing
+
+Second test is ensuring that Cincinnati instance can handle the load in order to adhere to SLO requirements from app-sre team. Load-testing is started by [load-testing.sh](../../hack/load-testing.sh) script.
+This script would start multiple connections to deployed Cincinnati instance using [vegeta](https://github.com/tsenart/vegeta) tool. The tool would check that repeated requests to [targets](../../hack/vegeta.targets) are return HTTP 2xx code. Vegeta would vary amount of parallel connections (`workers` in `load-testing.sh`) and rate of requests (`rate` in `load-testing.sh`). The tool would also ensure that the response is received in specified `duration` (30s).
```
Comment (steveej) [inline docs/developer/end-to-end-testing.md]:
```suggestion
This script starts multiple connections to the deployed Cincinnati instance using [vegeta](https://github.com/tsenart/vegeta) tool. The tool checks that repeated requests to [targets](../../hack/vegeta.targets) return a HTTP 2xx code. Vegeta varies the amount of parallel connections (`workers` in `load-testing.sh`) and rate of requests (`rate` in `load-testing.sh`). The tool also ensures that the response is received within the specified `duration` (30s).
```

File: docs/developer/end-to-end-testing.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+### SLO verification
+
+Once the load has been put on Cincinnati the third test - `check_slo` from [slo.rs](../../e2e/tests/slo.rs) - would ensure that metric values didn't break established SLO. The test queries user workload Prometheus, runs  queries specified as test case parameters and ensures expected result is received. SLO test ensures that neither service stopped reporting, container didn't restart, images have been scraped without errors, and any request has been processed in less that 0.5 second.
```
Comment (steveej) [inline docs/developer/end-to-end-testing.md]:
```suggestion
Once the load has finished the third test - `check_slo` from [slo.rs](../../e2e/tests/slo.rs) - ensures that metric values didn't break the established SLO. The test queries user workload Prometheus, runs queries specified as test case parameters and ensures expected result is received. SLO test ensures that neither service stopped reporting, container didn't restart, images have been scraped without errors, and any request has been processed in less that 0.5 second.
```

File: docs/developer/end-to-end-testing.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+If any of these tests fails `e2e-aws` task would fail and not proceed further
```
Comment (steveej) [inline docs/developer/end-to-end-testing.md]:
```suggestion
If any of these tests fail, the `e2e-aws` task fails and doesn't proceed further
```

**Review by steveej — COMMENTED** [review]
Thanks for writing this up! It's useful to have a documentation of how the e2e tests, especially the load test, are set up.

File: docs/developer/end-to-end-testing.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+Additionally it reconfigures the cluster's Prometheus cluster to support [user workloads](https://docs.openshift.com/container-platform/4.4/monitoring/monitoring-your-own-services.html). This spawns a new 
+Prometheus instance, which is used to scrape Cincinnati's metrics (using [observability](../../dist/openshift/observability.yaml) template).
+
+## end to end tests
+
+### Graph-builder and policy-engine functionality
+
+Once the script is done it starts the end-to-end tests from [e2e package](../e2e/tests). 
+
+First test (`e2e_channel_success`) checking the graph in the deployed Cincinnati and ensure that [expected releases](../../e2e/test/testdata) are present in the graph and have valid metadata. The test verifies that graph-builder is able to scrape production images from Quay correctly and policy-engine properly filters the graph by channel and architecture.
```
Comment (steveej) [inline docs/developer/end-to-end-testing.md]: The link to the testdata is dead

File: docs/developer/end-to-end-testing.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+* wait for Cincinnati to rollout, expecting liveness/readiness probes to pass
+
+Additionally it reconfigures the cluster's Prometheus cluster to support [user workloads](https://docs.openshift.com/container-platform/4.4/monitoring/monitoring-your-own-services.html). This spawns a new 
+Prometheus instance, which is used to scrape Cincinnati's metrics (using [observability](../../dist/openshift/observability.yaml) template).
+
+## end to end tests
+
+### Graph-builder and policy-engine functionality
+
+Once the script is done it starts the end-to-end tests from [e2e package](../e2e/tests).
```
Comment (steveej) [inline docs/developer/end-to-end-testing.md]: dead link

**Review by steveej — COMMENTED** [review]
Found some dead links. After that I think it can be squashed and merged.

File: docs/developer/end-to-end-testing.md (hunk 5 above)
Comment (vrutkovs) [reply docs/developer/end-to-end-testing.md]: Fixed in 0b9737f

**Review by steveej — APPROVED** [review]
This looks good now, thanks! Would you like to squash the commits or merge as is?

**Comment by vrutkovs** [comment]
Squashed commits, ptal

**Review by steveej — APPROVED** [review]
/lgtm

**Pull request merged**